### PR TITLE
Refactor Permission structure in Decision API

### PR DESCRIPTION
**Pull request by meravingen123**
Change Permission from pair to a struct and
add missing resource field.

Resolves: OLPSUP-10514

Signed-off-by: Serhii Lozynskyi <ext-serhii.lozynskyi@here.com>

### CODE REVIEW

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 1]
```diff
@@ -42,9 +42,17 @@ enum class DecisionType { kAllow, kDeny };
 class AUTHENTICATION_API ActionResult {
  public:
   /**
-   * @brief Represents the permission pair with the action and policy decision.
+   * @brief Represents the permission with the action, policy decision and
+   * associated resource.
    */
-  using Permissions = std::pair<std::string, DecisionType>;
+  struct Permission {
```
Comment (andescu) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:48]: This breaks a little our normal flow. The normal flow is that these structs look like swagger generated.
This includes getters/setter for all members and a separate file.

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 2]
```diff
@@ -19,11 +19,11 @@
 
 #pragma once
 
+#include <olp/authentication/AuthenticationApi.h>
```
Comment (mykhailo-kuchma) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:22]: Reorder the includes

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 3]
```diff
@@ -32,6 +32,60 @@ namespace authentication {
...
+   * @brief Sets the decision associated with the resource
+   *
+   * @param Decision to associate with the resource
+   */
+  void SetDecision(DecisionType decision) { decision_ = decision; }
+
+  /**
+   * @brief Gets the decision associated with the resource.
+   *
+   * @return Decision for the associated resource.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:79]:
```suggestion
   * @return The decision for the associated resource.
```

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 4]
```diff
@@ -32,6 +32,60 @@ namespace authentication {
...
+   * @brief Gets the resource with which action and decision are associated.
+   *
+   * @return Resource name.
+   */
+  const std::string& GetResource() const { return resource_; }
+
+  /**
+   * @brief Sets the decision associated with the resource
+   *
+   * @param Decision to associate with the resource
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:72]:
```suggestion
   * @brief Sets the decision associated with the resource.
   *
   * @param The decision to associate with the resource.
```

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 5]
```diff
@@ -32,6 +32,60 @@ namespace authentication {
...
+   * @brief Sets the resource with which action and decision are associated
+   *
+   * @param Resource to associate with decision and action
+   */
+  void SetResource(std::string resource) { resource_ = std::move(resource); }
+
+  /**
+   * @brief Gets the resource with which action and decision are associated.
+   *
+   * @return Resource name.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:65]:
```suggestion
   * @brief Gets the resource with which the action and decision are associated.
   *
   * @return The resource name.
```

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 6]
```diff
@@ -32,6 +32,60 @@ namespace authentication {
...
+   * @brief Gets the action that is associated with the resource
+   *
+   * @return String representing action
+   */
+  const std::string& GetAction() const { return action_; }
+
+  /**
+   * @brief Sets the resource with which action and decision are associated
+   *
+   * @param Resource to associate with decision and action
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:58]:
```suggestion
   * @brief Sets the resource with which the action and decision are associated.
   *
   * @param The resource to associate with decision and action.
```

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 7]
```diff
@@ -32,6 +32,60 @@ namespace authentication {
...
+   * @brief Sets the action associated with the resource
+   *
+   * @param The action to associate with.
+   */
+  void SetAction(std::string action) { action_ = std::move(action); }
+
+  /**
+   * @brief Gets the action that is associated with the resource
+   *
+   * @return String representing action
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:51]:
```suggestion
   * @brief Gets the action that is associated with the resource.
   *
   * @return A string that represents the action.
```

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 8]
```diff
@@ -32,6 +32,60 @@ namespace authentication {
...
 enum class DecisionType { kAllow, kDeny };
 
+/**
+ * @brief Represents the permission with the action, policy decision and
+ * associated resource.
+ */
+class AUTHENTICATION_API Permission {
+ public:
+  /**
+   * @brief Sets the action associated with the resource
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:42]:
```suggestion
   * @brief Sets the action associated with the resource.
```

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 9]
```diff
@@ -32,6 +32,60 @@ namespace authentication {
  */
 enum class DecisionType { kAllow, kDeny };
 
+/**
+ * @brief Represents the permission with the action, policy decision and
+ * associated resource.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:37]:
```suggestion
 * @brief Represents the permission with the action, policy decision, and
 * associated resource.
```

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 10]
```diff
@@ -32,6 +32,60 @@ namespace authentication {
...
+
+  /**
+   * @brief Gets the action that is associated with the resource.
+   *
+   * @return A string that represents the action.
+   */
+  const std::string& GetAction() const { return action_; }
+
+  /**
+   * @brief Sets the resource with which action and decision are associated.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:56]:
```suggestion
   * @brief Sets the resource with which the action and decision are associated.
```

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h [hunk 11]
```diff
@@ -32,6 +32,60 @@ namespace authentication {
...
+   *
+   * @param The resource to associate with decision and action.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthorizeResult.h:58]:
```suggestion
   * @param The resource to associate with the decision and action.
```

File: olp-cpp-sdk-authentication/src/AuthenticationClientUtils.cpp [hunk 12]
```diff
@@ -177,18 +177,22 @@ std::vector<auth::ActionResult> GetDiagnostics(rapidjson::Document& doc) {
...
           if (permission_element.HasMember(Constants::ACTION)) {
-            permission.first =
-                permission_element[Constants::ACTION].GetString();
+            permission.SetAction(
+                permission_element[Constants::ACTION].GetString());
           }
           if (permission_element.HasMember(Constants::DECISION)) {
-            permission.second = GetPermission(
-                permission_element[Constants::DECISION].GetString());
+            permission.SetDecision(GetPermission(
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/AuthenticationClientUtils.cpp:190]: Maybe `GetPermission` is not the best name. Should be `GetDecision`.

**Pull request merged**